napari/plugins/_npe2.py

```diff
@@ -282,11 +294,115 @@ def _on_plugin_enablement_change(enabled: Set[str], disabled: Set[str]):
     to_disable.update(disabled)
     plugin_settings.disabled_plugins = to_disable
 
+    for plugin_name in enabled:
+        # technically, you can enable (i.e. "undisable") a plugin that isn't
+        # currently registered/available.  So we check to make sure this is
+        # actually a registered plugin.
+        if plugin_name in pm.instance():
+            _register_manifest_actions(pm.get_manifest(plugin_name))
+
+    # TODO: after app-model, these QMenus will be evented and self-updating
+    # and we can remove this... but `_register_manifest_actions` will need to
+    # add the actions file and plugins menus (since we don't require plugins to
+    # list them explicitly)
     for v in Viewer._instances:
         v.window.plugins_menu._build()
         v.window.file_menu._rebuild_samples_menu()
 
 
-def index_npe1_adapters():
-    """Tell npe2 to import and index any discovered npe1 plugins."""
-    pm.index_npe1_adapters()
+def on_plugins_registered(manifests: Set[PluginManifest]):
+    """Callback when any npe2 plugins are registered.
+
+    'Registered' means that a manifest has been provided or discovered.
+    """
+    for mf in manifests:
+        if not pm.is_disabled(mf.name):
+            _register_manifest_actions(mf)
+
+
+def _register_manifest_actions(manifest: PluginManifest) -> None:
+    """Gather and register actions from a manifest.
+
+    This is called when a plugin is registered or enabled and it adds the
+    plugin's menus and submenus to the app model registry.
+    """
+    from .._app import app
+
+    actions, submenus = _npe2_manifest_to_actions(manifest)
+    if actions:
+        disposable = app.register_actions(actions)
+        pm.get_context(manifest.name).register_disposable(disposable)
+    if submenus:
+        disposable = app.menus.append_menu_items(submenus)
+        pm.get_context(manifest.name).register_disposable(disposable)
+
+
+def _is_menu_contributable(menu_id) -> bool:
+    """Return True if the given menu_id is a menu that plugins can contribute to."""
+    # here is where we can check whether the menu id is "contributable"
+    # or not.  i.e. if they're trying to add to the File menu, we skip it here.
+    from .._app.constants import MenuId
+
+    if menu_id.startswith("napari/"):
+        # emit a warning?
+        return menu_id in {MenuId.LAYERLIST_CONTEXT.value}
```

Comment:
is this currently the 'single source of truth' for contributable menus? Would it be nicer to have an Enum containing these somewhere which we reference here?

Reply:
yes, and yes :) 
we need something we can easily mine for docs at napari.org.  will do in this PR